[AMBARI-23595] Generic method to replace HOSTGROUP variable in blueprints

benyoka opened the pull request:
## What changes were proposed in this pull request?

This change is to give more flexibility for using ```%HOSTGROUP::name%``` placeholders in blueprints. The following changes were made:
- Refactored ```doUpdateForClusterCreate()``` in ```BlueprintConfigurationProcessor``` for better readablility. General and Namenode HA property updates were both extracted to their own methods, ```doGeneralPropertyUpdatesForClusterCreate()``` and ```doNameNodeHAUpdateOnClusterCreation()``` respectively. 
- In ```doNameNodeHAUpdateOnClusterCreation()``` in addition to the existing update logic, a default updater is called on all properties defined in the blueprint and cluster template for which a registered updater doesn't exist to make sure ```%HOSTGROUP``` tokens are replaced. The default updater is a ```HostGroupupdater```. The default updater is not called on stack default properties.
- Updated the logic in ```HostGroupupdater```. Previously, only the first occurence of a %HOSTGROUP token would be replaced in a property. E.g. the following property was not handled correctly: ``` "registry.url": "http://%HOSTGROUP::master_2%:7788/api/v1,http://%HOSTGROUP::master_3%:7788/api/v1"```. The new logic replaces all occurences of ```%HOSTGROUP``` in property values the following way:
   - Multiple host groups can be referenced in a value, like the example above
   - The same host group can occur multiple times. For each occurrence another host is picked from the host group. (If all hosts are consumed, it will start over with the first host. This can be easily changed to throw an exception.) 
   - This gives some flexibility without introducing new syntax. E.g. creating a property value that lists all hosts in a host group is now possible assuming that the number of host is known beforehand, e.g: ```%HOSTGROUP::master1%:8080,%HOSTGROUP::master1%:8080```

## How was this patch tested?
- Tested manually with HDF blueprints. Deployment didn't succeed due to unrelated reasons, but property replacement happened as expected.
- Wrote new unit tests
- Unit test results:
```
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   BufferedThreadPoolExecutorCompletionServiceTest.testScalingThreadPoolExecutor:181 expected:<10> but was:<9>
[ERROR] Errors:
[ERROR]   CredentialStoreTest.testInMemoryCredentialStoreService_CredentialExpired:90->getExpiredCredentialTest:169 NullPointer
[INFO]
[ERROR] Tests run: 5088, Failures: 1, Errors: 1, Skipped: 68
```
Both failed tests are unrelated and were successfully re-run.

Commit:
AMBARI-23595 add default updater to replace %HOSTGROUP tokens (benyoka)

Commit:
AMBARI-23595 changed the algorithm a bit (benyoka)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/1952/
Test PASSed.

adoroszlai commented on ambari-server/src/main/java/org/apache/ambari/server/controller/internal/BlueprintConfigurationProcessor.java:
```diff
@@ -359,152 +362,245 @@ private boolean containsHostFromHostGroups(String configType, String propertyNam
+   * @param clusterConfig
+   * @param clusterProps
+   * @param configTypesUpdated
+   * @throws ConfigurationTopologyException
```
Please don't add such javadoc tags without description (as it doesn't add any value to the information already available from the signature).  Please either add description, or drop the tag.

adoroszlai commented on ambari-server/src/main/java/org/apache/ambari/server/controller/internal/BlueprintConfigurationProcessor.java:
```diff
@@ -359,152 +362,245 @@ private boolean containsHostFromHostGroups(String configType, String propertyNam
+   * @param propertyName the name of the property (e.g. dfs.namenode.name.dir)
+   * @param oldValue the old value of the property
+   * @param updater the property updater to use
+   * @param allProps other properties to be considered by the upsater
```
typo: `upsater -> updater`

benyoka replied:
Thanks

adoroszlai reviewed (approved):
LGTM.  Two minor nits below.

rnettleton commented on ambari-server/src/main/java/org/apache/ambari/server/controller/internal/BlueprintConfigurationProcessor.java:
```diff
@@ -359,152 +362,245 @@ private boolean containsHostFromHostGroups(String configType, String propertyNam
+            String propertyName = propertyEntry.getKey();
+            if (!propertiesWithUpdaters.contains(Pair.of(configType, propertyName))) {
+              String oldValue = propertyEntry.getValue();
+              updateValue(configType, propertyName, oldValue, HostGroupUpdater.INSTANCE, properties, configuration, configTypesUpdated, false);
```
It might be a good idea to add some logging here, to indicate that default HOSTGROUP token matching is taking place. This would be very helpful in the event that a user needs to debug a Blueprint config processing result that doesn't match the expected output.

benyoka replied:
Ok.

rnettleton commented on ambari-server/src/main/java/org/apache/ambari/server/controller/internal/BlueprintConfigurationProcessor.java:
```diff
@@ -1472,31 +1568,35 @@ String updateForClusterCreate(String propertyName,
+      //      group3_host1:8080,group3_host2:8080,group3_host3:8080 (maybe in different order)
+      for (Matcher m = HostGroup.HOSTGROUP_REGEX.matcher(origValue); m.find(); ) {
+        origValue = m.replaceFirst(hostGroups.nextHost(m.group(1)));
+        m = HostGroup.HOSTGROUP_REGEX.matcher(origValue);
```
Is there a possibility that this could result in an infinite loop?  

For example: If a HOSTGROUP token existed in a property that did not match to one of the host group names defined in the Blueprint, then "origValue" might never get updated, causing the regex matcher to keep detecting the HOSTGROUP token that was never replaced.  

Can you please investigate this, to make sure this isn't a possibility?  If it is, then maybe change the loop conditions to guarantee that an infinite loop won't occur in the even of a mis-configured Blueprint.

benyoka replied:
Good catch. It actually could result in an infinite loop in the unlikely case when a host group is named "%HOSTGROUP::something%. I rewrote the replacement logic to take care of that.

rnettleton reviewed (commented):
Overall, the patch looks fine to me, but I think there are a few things to consider prior to merging this: 

1. I think adding support for multiple occurrences of a hostgroup name in a comma-separated list might not be something we should introduce in Ambari 2.7.0.  I'm concerned that this behavior might be confusing to users of existing Blueprints.  In addition, I'm not sure it would help much in Ambari 2.7.0, since multiple instances of a given service can't be configured to run on the same host.  This support is something we should consider adding in Ambari 3.0, but in Ambari 2.7.0 I think we should throw an exception when this situation occurs. 

2. I would ask that you test out multiple deployment scenarios with Blueprints (HDFS HA, non-HA deployments, NameNode Federation), to ensure that these changes to the BlueprintConfigurationProcessor don't have any unintended side effects.  

Thanks for adding this patch, since the new HOSTGROUP token matching for properties that do not have an updater registered will be very helpful for new services added to a given stack.

benyoka commented:
Hey @rnettleton,

In your 2. point do you mean:
1. Removing the cyclic iterator, so **this will be possible:**
```"myservice.slave.hosts": "%HOSTGROUP::slaves_1%,%HOSTGROUP::slaves_1%"``` (assuming the slaves_1 host group has at least 2 hosts)
**but this one will not:** 
```""myservice.instances": "%HOSTGROUP::master_1%:8080,%HOSTGROUP::master_1%:8080"``` (assuming master_1 has only one host)
2. Completely removing the support for referencing the same host group multiple times.

Commit:
AMBARI-23595 fix review comments + update getRequiredHostGroups (benyoka)

benyoka commented:
@rnettleton, @adoroszlai  I have updated the PR with the review comments and also updated getRequiredHostgroups() - I forgot this in the original PR.

@rnettleton please let me know if removing the cyclic iterator and throwing an exception in case a host group is referenced more times in a property than the number of host in that group is the change you wanted to make.

Commit:
AMBARI-23595 some minor fixes (benyoka)

rnettleton commented on ambari-server/src/main/java/org/apache/ambari/server/controller/internal/BlueprintConfigurationProcessor.java:
```diff
@@ -1470,36 +1619,59 @@ String updateForClusterCreate(String propertyName,
+      // the following replacements will be made:
+      // - %HOSTGROUP::group1%:8080 --> grop1_host:8080
+      // - %HOSTGROUP::group1%:8080,%HOSTGROUP::group2%:8080 --> group1_host:8080,group2_host:8080
+      // - %HOSTGROUP::group3%:8080,%HOSTGROUP::group3%:8080,%HOSTGROUP::group3%:8080 -->
```
I think that support for a host_group being listed multiple times in a given property value should not be introduced in Ambari 2.7.0, so I would request that this support be removed from this PR.  

I do think there might be some value in adding something like this in Ambari 3.0, when the multi-* support is added, but in 2.7.0 I think the addition would be too confusing.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/1994/
Test PASSed.

rnettleton commented on ambari-server/src/test/java/org/apache/ambari/server/controller/internal/BlueprintConfigurationProcessorTest.java:
```diff
@@ -2168,6 +2177,198 @@ public void testDoUpdateForClusterCreate_SingleHostProperty__defaultValue() thro
+      topology);
+  }
+
+  @Test(expected = IllegalStateException.class)
```
Minor issue:  Where is this check occurring, and where is this exception thrown?  I've looked through this patch a few times, but I wasn't able to determine where the check for multiple instances of a given hostgroup name occurring in a property value.  

I'm probably just missing something, so if you could clarify this it would be a big help.

rnettleton reviewed (approved):
Overall, the changes look fine to me, just a few minor issues that need to be handled/clarified prior to merging. 

Thanks again for creating this patch!

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/1995/
Test PASSed.

Commit:
AMBARI-23595 more review fixes (benyoka)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/2006/
Test FAILed.
Test FAILured.

benyoka commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/2009/
Test PASSed.

Commit:
Merge remote-tracking branch 'upstream/trunk' into AMBARI-23595-trunk

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/2132/
Test FAILed.
Test FAILured.

benyoka commented:
Hey @adoroszlai , I merged trunk into this feature branch and as part of it I had to merge your changes to BlueprintConfigurationProcessor by hand. Could you please check the changes to BlueprintConfigurationProcessor in the latest commit.

cc @rnettleton

benyoka commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/2138/
Test PASSed.

adoroszlai commented:
@benyoka, the merge looks good.
